detectron2/modeling/anchor_generator.py

```diff
@@ -36,13 +36,17 @@ def __iter__(self):
         return iter(self._buffers.values())
 
 
-def _create_grid_offsets(size: List[int], stride: int, offset: float, device: torch.device):
+def _create_grid_offsets(
+    size: List[int], stride: int, offset: float, target_device_tensor: torch.Tensor
+):
     grid_height, grid_width = size
-    shifts_x = torch.arange(
-        offset * stride, grid_width * stride, step=stride, dtype=torch.float32, device=device
+    shifts_x = move_device_like(
+        torch.arange(offset * stride, grid_width * stride, step=stride, dtype=torch.float32),
```

Comment:
Here it also uses `arange` so it has the same issue: `arange` is created on old device and moved to new device, instead of created on the new device directly.

Reply:
To find all problems like this, maybe consider a unittest that traces the model on GPU, and then run the traced model on CPU in a separate process whose `CUDA_VISIBLE_DEVICES=""`. Such a test is supposed to fail for this line (unless pytorch has some smart optimizations).

Reply:
My impression is for all torch's functions that doesn't take any tensor as input (eg. `torch.arange`) without specifying the device, it'll always be on `cpu` even when running on GPU device. So this code should work for tracing on one GPU and running on another GPU or CPU. Nonetheless, I'm gonna write some GPU tests covering this.

Performance wise I'm not sure how large is the difference between creating one on CPU then moving to GPU versus creating on GPU directly. Suppose creating tensor on CPU is negligible.